### ISSUE TITLE
Make a demo of device speaking ($1500)

### ISSUE THREAD

**Issue by kodjima33** [BasedHardware/omi#1006]
**Is your feature request related to a problem? Please describe.**
People want to talk to the device. Our v2 device is equipped with a speaker but doesn't speak yet 

**Describe the solution you'd like**
Solution should be exactly like our in-app chat but with audio from device. For example, I click on the button (on device) and ask "hey what's the  capital of united states" and device should respond with audio "Washington DC". This response and prompt should be visible on the chat inside of the app

This functionality should be disabable via settings. 

**Additional context**
There was a PR submitted a while ago to make the app speak, take a look at that 

This is a paid task. Reward is $1,500 in cash. Simply link your PR to this issue to receive the money

**Comment by beastoin** [comment]
sweet!

**Comment by DamienDeepgram** [comment]
I can get this working but would be using Deepgram only.

**Comment by beastoin** [comment]
Go a head with a PR pls @DamienDeepgram , you don't need a permission to do great things.

**Comment by beastoin** [comment]
@hoai265 you too - great firmware dev man

**Comment by kodjima33** [comment]
Sorry guys had to update the bounty

**Comment by beastoin** [comment]
so sad, no one take this :dead: too sweet $1K

**Comment by Sanchay-T** [comment]
Hi @kodjima33 , I’d love to take this task. Please assign it to me, and I’ll get started. Thanks!

**Comment by beastoin** [comment]
@Sanchay-T yes, pls keep us updated! every 24h would be great ~ just to keep your motivation up!

**Comment by Sanchay-T** [comment]
Hi @beastoin 👋 

I've been diving into the codebase to understand how we can implement voice responses for the device. Really interesting architecture you've built here! While I'm familiar with backend systems, I'm getting up to speed with some of the Flutter and BLE specifics.

Looking at the current audio pipeline, I can see we're handling real-time streaming through WebSockets pretty elegantly. The socket service in `app/lib/services/sockets.dart` seems to be the core of this:

```dart
Future<TranscriptSegmentSocketService?> socket({
    required BleAudioCodec codec,
    required int sampleRate,
    required String language,
    bool force = false,
}) async {
```

For implementing the voice response feature, I think we can build on this foundation. I see we're already integrated with OpenAI's APIs in `backend/http/openai.dart`, which could be extended for text-to-speech capabilities.

I have a few questions about the device interaction part though. Looking at `app/lib/services/devices/models.dart`, I see how we're handling BLE characteristics:

```dart
BluetoothCharacteristic? getCharacteristicByUuid(BluetoothService service, String uuid) {
  return service.characteristics.firstWhereOrNull(
    (characteristic) => characteristic.uuid.str128.toLowerCase() == uuid.toLowerCase(),
  );
}
```

Before I proceed with the implementation, I wanted to check:

1. For handling button presses - what would be the best way to detect when the user wants to trigger a voice command? Should we use an existing characteristic or define a new one?

2. Regarding audio playback through the device's speaker - are there any specific format requirements or limitations I should be aware of?

3. For the chat interface, I see we're using the Memory system to handle conversations. Would adding voice responses require any significant changes to the current schema?

I have some ideas about the implementation, but wanted to validate these core aspects first to make sure I'm heading in the right direction. Happy to elaborate on any part of this!

Thanks for the help! Looking forward to your insights.

**Comment by beastoin** [comment]
1/ what do your propose ? pros / cons.
2/ @kevvz could help ? but you should try it yourself first.
3/ just do it (to know that you're wrong 😏)
no worries man, be creative. let's finish the first draft quickly then we have something to discuss. embracing the changes( good changes :))

@Sanchay-T

**Comment by DamienDeepgram** [comment]
Deepgram also has TTS so you could use the same sdk I think that the speech to text is using. Not sure if Omi has a preference there tho

see: https://pub.dev/packages/deepgram_speech_to_text#text-to-speech

**Comment by kodjima33** [comment]
Removed @Sanchay-T from assigned - no progress

@beastoin let's try to not assign people if they didn't yet have PRs submitted. We assign only to those who had PRs. Others will need to do a PR first. 

@DamienDeepgram try it out bro - looking forward!

**Comment by Sanchay-T** [comment]
Hi @beastoin and @kodjima33 

I wanted to clarify the situation regarding my previous assignment. First, I apologize for the delay in updates - I was away for Diwali celebrations in my hometown, which affected my response time. However, I want to assure you that I've been actively working on this in the background:

1. I've been going through the codebase thoroughly, particularly focusing on the audio pipeline and BLE integration
2. While I have less experience with Flutter/Dart specifically, I bring relevant experience with speech/text models which I believe will be valuable for this feature
3. I'm currently working on implementing a proof-of-concept to address the questions I raised earlier, particularly around:
   - Button press handling for voice command triggering
   - Audio playback implementation
   - Memory system integration for voice responses

I understand the policy about assignments and PRs, and I'm committed to submitting a PR with my implementation soon. I should have communicated my temporary absence better, and I appreciate your patience.

Would it be alright if I continue working on this feature and submit a PR for review? I'm happy to share my current progress in more detail if helpful.

Thanks for understanding!

**Comment by beastoin** [comment]
checking speaker functions of current firmware...

**Comment by hoai265** [comment]
> Removed @Sanchay-T from assigned - no progress
> 
> @beastoin let's try to not assign people if they didn't yet have PRs submitted. We assign only to those who had PRs. Others will need to do a PR first.
> 
> @DamienDeepgram try it out bro - looking forward!

Hi @kodjima33 What if we add a separate option to route audio to the phone's output, like AirPods? I think this would be another option for users, as they could listen privately.

**Comment by beastoin** [comment]
@DamienDeepgram don't forget to ref your PR ;) 

about the preference to implement this task, be creative. but i think Nik's description is good/simple enough to roll out l the first draft. smth likes ~ 

1/ the user press the button in the device and say something
2/ the device send that voice to the app
3/ the app send the voice message to the backend
4/ the backend process the voice message then response to the app with audio bytes
5/ the app send the audio bytes to the device
6/ the device speak it out loud.

hope that helps.

**Comment by vincentkoc** [comment]
speaker should support playback over BT, i dont might looking into this after apple watch PR as it will use a similar two-way transport.

**Comment by vincentkoc** [comment]
I have started on this, but need to get some sleep #1243 - i think even before i flashed new firmware any button click on my (red) devkit2 causes a fatal crash - not sure if the shipped devkit2's have a different setup with button? Could be a different pin/setup for the button causing this issue.

Code in WIP includes all the BLE setup to stream and handle the stream on desktop side (Python). Once finalised can move to dart code.

**Comment by kodjima33** [comment]
@Sanchay-T bro no worries, just keep building this and try to make it work. No one blames you - it's just we assign issues only after first PR 

@vincentkoc @DamienDeepgram guys I believe in you. Let's make this work! (ideally today)

You are both working on this, if you both make it work, I'll make a post about both of you and we will solve the bounty issue

**Comment by beastoin** [comment]
fighting 💪

**Comment by DamienDeepgram** [comment]
> @DamienDeepgram don't forget to ref your PR ;)

Sorry yes here is the PR with the issue with playback not streaming correctly

https://github.com/BasedHardware/omi/pull/1246

**Comment by ombhojane** [comment]
Hey @kodjima33 @beastoin 
Is this issue been resolved? I've gone through discussions and am willing to develop this, may I proceed?

**Comment by beastoin** [comment]
Please go a head and keep us updated @ombhojane

**Comment by ombhojane** [comment]
Sure @beastoin

**Comment by ombhojane** [comment]
@beastoin 
I'm stucked at setup the project.
With Omi's instructions, I did setup, at last stage it was building android gradle files, it downloaded more data than expected. 
So more time was gone in setting up the things, and still figuring out. 
Need to see what's going on and how to fix.

**Comment by himmat12** [comment]
So how do we integrate this feature if we do not have Omi dev kit devices? Or, can we implement this feature in our Android/iOS device, and if it works there, then it works with OMI devices?

**Comment by beastoin** [comment]
@himmat12 you already know the answer man. the ticket title is super clear.

**Comment by beastoin** [comment]
@ombhojane how's it going? 

if you want to get this tiket done - building the app / the firmware is a basic requirement.

**Comment by ombhojane** [comment]
Hey @beastoin I'll figure this out today, yesterday was my exam.
I'll try manual installation once.

**Comment by ombhojane** [comment]
Hii @beastoin 
I've set up the Omi. Now I'm looking to fix the issue, I'll update the progress

**Comment by vincentkoc** [comment]
@beastoin @kodjima33 is this fixed with #1452 or work still required?

**Comment by kodjima33** [comment]
I haven't seen any working demo for this yet

**Comment by vincentkoc** [comment]
Just got the DK2 on weekend. Starting working on this again this week.Sent from my iPhoneOn Feb 4, 2025, at 7:59, Nik Shevchenko ***@***.***> wrote:﻿
I haven't seen any working demo for this yet

**Comment by andrewgazelka** [comment]
@addbounty $1500

**Comment by amanem1** [comment]
hi guys is this issue still open or it is assigned to anyone specific , just wanted to know the status of the issue as to continue working on it or to start another issue ? 
@andrewgazelka @kodjima33

**Comment by addbounty** [comment]
[![banner](https://bountybot.dev/api/display/github/BasedHardware/omi/1006/banner?v=1a36cfb6-24c4-422b-9258-fb1d505243ed)](https://bountybot.dev/github/BasedHardware/omi/issues/1006)
[![button](https://bountybot.dev/api/display/github/BasedHardware/omi/1006/button?v=1a36cfb6-24c4-422b-9258-fb1d505243ed)](https://bountybot.dev/github/BasedHardware/omi/issues/1006)

Make a Draft PR **early** so others can see you are working on it! To automatically create one:

```sh
# Using npx (installed if you have NodeJS/npm)
npx bountybot start BasedHardware/omi#1006

# Or, use cargo (installed if you have Rust)
cargo install bounty
bounty start BasedHardware/omi#1006
```

When merged, you will receive the bounty!

**Comment by andrewgazelka** [comment]
> hi guys is this issue still open or it is assigned to anyone specific , just wanted to know the status of the issue as to continue working on it or to start another issue ? [@andrewgazelka](https://github.com/andrewgazelka) [@kodjima33](https://github.com/kodjima33)

I would wait on working on this until `kodjima33` responds as there is an existing assignee.

Note: it is not listed as locked but this might be an issue.

<img width="899" alt="Image" src="https://github.com/user-attachments/assets/9462d7c2-b75f-4587-9a95-3f9444b46289" />

**Comment by kodjima33** [comment]
I haven't yet see any working demos for this, meaning that anyone can pick this up - issue is open

However I do know that @vincentkoc is trying very hard

**Comment by vincentkoc** [comment]
I have pushed up my branch so everyone can see my progress. Started a new branch a few weeks ago. I had issues with button so new strategy is:

- Control signal to switch the device from [mic listen] and write to [mic off and read to play speaker]
- Using a test script to simulate client app

Once above is confirmed and device can "speak", this demo will be switched to deepgram (i have a local version of the test script with this to generate the wav to send) and perhaps a two way with the button if I get time.

**Comment by kodjima33** [comment]
seems impossible with current speaker quality -removing
